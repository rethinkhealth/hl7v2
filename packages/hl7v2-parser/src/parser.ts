import type { Root } from '@rethinkhealth/hl7v2-ast';
import { DEFAULT_DELIMITERS } from '@rethinkhealth/hl7v2-utils';
import type { Plugin } from 'unified';
import { defaultPreprocessors, runPreprocessors } from './preprocessor';
import { parseHL7v2FromIterator } from './processor';
import { HL7v2Tokenizer } from './tokenizer';
import type { ParseOptions, ParserContext, Token, Tokenizer } from './types';

function* iterateTokenizerSync(t: Tokenizer): Iterable<Token> {
  for (let tok = t.next(); tok; tok = t.next()) {
    yield tok;
  }
}

export function parseHL7v2(input: string, opts: ParseOptions): Root {
  let ctx: ParserContext = {
    input,
    delimiters: opts.delimiters || DEFAULT_DELIMITERS,
  };
  // Run preprocessing
  ctx = runPreprocessors(ctx, opts.preprocess || defaultPreprocessors);
  // Run tokenizer
  const tokenizer = new HL7v2Tokenizer();
  // Reset tokenizer.
  tokenizer.reset(ctx);
  // Parse
  return parseHL7v2FromIterator(iterateTokenizerSync(tokenizer), ctx);
}

const hl7v2Parser: Plugin<[ParseOptions?], string, Root> = function (
  options = {}
) {
  this.parser = (document: string) => parseHL7v2(document, options);
};

export default hl7v2Parser;
