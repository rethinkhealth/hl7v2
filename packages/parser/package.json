{
  "name": "@rethinkhealth/hl7v2",
  "description": "Modern Typescript HL7v2 package",
  "version": "0.1.0",
  "license": "MIT",
  "author": {
    "name": "Melek Somai",
    "email": "melek@rethinkhealth.io"
  },
  "type": "module",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": "./dist/index.js"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "compile": "tsc --noEmit",
    "build": "tsup && tsc --emitDeclarationOnly"
  },
  "devDependencies": {
    "@rethinkhealth/tsconfig": "workspace:*",
    "@biomejs/biome": "1.9.4",
    "@changesets/cli": "2.29.4",
    "@types/node": "22.15.31",
    "syncpack": "13.0.4",
    "tsup": "8.5.0",
    "typescript": "5.8.3"
  },
  "repository": "rethinkhealth/hl7v2.git",
  "homepage": "https://www.rethinkhealth.io/hl7v2/docs",
  "keywords": ["health", "healthcare", "hl7", "hl7v2", "nodejs", "typescript"],
  "packageManager": "pnpm@10.12.1",
  "publishConfig": {
    "access": "public"
  }
}
