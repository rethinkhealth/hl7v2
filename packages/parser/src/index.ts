export { HL7v2Client } from "./client";
export { DefaultDelimiters } from "./delimiters";
export type { IDelimiters } from "./delimiters";
