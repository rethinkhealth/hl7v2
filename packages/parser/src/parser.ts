import { DEFAULT_DELIMITERS, EMPTY_MESSAGE } from './constants';
import type { HL7v2Delimiters, HL7v2Node } from './types';
import { detectDelimitersFromMSH, splitByString } from './utils';

export interface ParseOptions {
  delimiters?: Partial<HL7v2Delimiters>;
  autoDetectDelimiters?: boolean;
}

/**
 * Parse an HL7v2 message into a Unist-compatible DOM tree.
 * Tracks lines/columns and preserves delimiters for round-tripping.
 */
export function parseHL7(
  rawMessage: string,
  options: ParseOptions = {}
): HL7v2Node {
  if (!rawMessage.trim()) {
    return EMPTY_MESSAGE;
  }

  // Merge user-defined delimiters with defaults
  const baseDelimiters: HL7v2Delimiters = {
    ...DEFAULT_DELIMITERS,
    ...options.delimiters,
  };

  // Auto-detect from MSH-2 if enabled
  let activeDelimiters = baseDelimiters;
  if (options.autoDetectDelimiters !== false) {
    const detected = detectDelimitersFromMSH(
      rawMessage,
      baseDelimiters.segment
    );
    activeDelimiters = { ...baseDelimiters, ...detected };
  }

  // Split into segments
  const segments = splitByString(rawMessage, activeDelimiters.segment);

  const messageNode: HL7v2Node = {
    type: 'message',
    children: [],
    position: {
      start: { line: 1, column: 1, offset: 0 },
      end: {
        line: segments.length,
        column: (segments.at(-1)?.value.length ?? 0) + 1,
        offset: rawMessage.length,
      },
    },
  };

  let currentLine = 1;
  for (let i = 0; i < segments.length; i++) {
    const seg = segments[i];

    if (!seg) {
      // TODO: Handle this case with a better error message
      throw new Error('Invalid message');
    }

    if (!seg.value.trim()) {
      currentLine++;
      continue;
    }

    const segmentNode = parseSegment(
      seg.value,
      seg.start,
      seg.end,
      currentLine,
      i,
      activeDelimiters
    );
    if (segmentNode) {
      if (!messageNode.children) {
        messageNode.children = [];
      }

      messageNode.children.push(segmentNode);
    }
    currentLine++;
  }

  return messageNode;
}

/**
 * Parse a single segment into fields and components
 */
function parseSegment(
  segmentText: string,
  segmentStart: number,
  segmentEnd: number,
  line: number,
  segmentIndex: number,
  delimiters: HL7v2Delimiters
): HL7v2Node | null {
  // 1. Split into fields
  const fields = splitByString(segmentText, delimiters.field);

  // 2. If no fields, return null
  if (!fields[0] || fields.length === 0) {
    return null;
  }

  const segmentNode: HL7v2Node = {
    type: 'segment',
    name: fields[0].value,
    index: segmentIndex,
    delimiter: delimiters.field,
    children: [],
    position: {
      start: { line, column: 1, offset: segmentStart },
      end: { line, column: segmentText.length + 1, offset: segmentEnd },
    },
  };

  for (let i = 0; i < fields.length; i++) {
    const f = fields[i];

    if (!f) {
      // TODO: Handle this case with a better error message
      throw new Error('Invalid message');
    }

    const fieldNode = createFieldNode(
      f.value,
      i,
      segmentStart + f.start,
      segmentStart + f.end,
      line,
      f.start + 1,
      delimiters
    );

    // biome-ignore lint/style/noNonNullAssertion: This is defined always
    segmentNode.children!.push(fieldNode);
  }

  return segmentNode;
}

/**
 * Create a field node, parsing components if needed
 */
function createFieldNode(
  fieldText: string,
  fieldIndex: number,
  fieldStart: number,
  fieldEnd: number,
  line: number,
  column: number,
  delimiters: HL7v2Delimiters
): HL7v2Node {
  const node: HL7v2Node = {
    type: 'field',
    index: fieldIndex,
    position: {
      start: { line, column, offset: fieldStart },
      end: { line, column: column + fieldText.length, offset: fieldEnd },
    },
  };

  if (fieldText.includes(delimiters.component)) {
    return {
      ...node,
      delimiter: delimiters.component,
      children: parseComponents(
        fieldText,
        fieldStart,
        line,
        column,
        delimiters
      ),
    };
  }

  return { ...node, value: fieldText };
}

/**
 * Parse components within a field
 */
function parseComponents(
  fieldText: string,
  fieldStart: number,
  line: number,
  column: number,
  delimiters: HL7v2Delimiters
): HL7v2Node[] {
  const comps = splitByString(fieldText, delimiters.component);
  const nodes: HL7v2Node[] = [];

  for (let i = 0; i < comps.length; i++) {
    const c = comps[i];

    if (!c) {
      // TODO: Handle this case with a better error message
      throw new Error('Invalid message');
    }

    const startColumn = column + c.start;
    const base: HL7v2Node = {
      type: 'component',
      index: i,
      position: {
        start: { line, column: startColumn, offset: fieldStart + c.start },
        end: {
          line,
          column: startColumn + c.value.length,
          offset: fieldStart + c.end,
        },
      },
    };

    if (c.value.includes(delimiters.subcomponent)) {
      nodes.push({
        ...base,
        delimiter: delimiters.subcomponent,
        children: parseSubcomponents(
          c.value,
          fieldStart + c.start,
          line,
          startColumn,
          delimiters
        ),
      });
    } else {
      nodes.push({ ...base, value: c.value });
    }
  }

  return nodes;
}

/**
 * Parse subcomponents within a component
 */
function parseSubcomponents(
  componentText: string,
  componentStart: number,
  line: number,
  column: number,
  delimiters: HL7v2Delimiters
): HL7v2Node[] {
  const subs = splitByString(componentText, delimiters.subcomponent);

  if (!subs[0]) {
    // TODO: Handle this case with a better error message
    throw new Error('Invalid message');
  }

  return subs.map((s, i) => ({
    type: 'subcomponent',
    index: i,
    value: s.value,
    position: {
      start: {
        line,
        column: column + s.start,
        offset: componentStart + s.start,
      },
      end: {
        line,
        column: column + s.start + s.value.length,
        offset: componentStart + s.end,
      },
    },
  }));
}
