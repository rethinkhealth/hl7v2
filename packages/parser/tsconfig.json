{
  "extends": "@rethinkhealth/tsconfig/library.json"
}