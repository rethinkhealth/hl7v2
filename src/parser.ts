import { DefaultDelimiters, IDelimiters } from "./delimiters";
import { ISegment, Segment } from "./segment";

export interface IParser {
  delimiters: IDelimiters;
  raw: string;
  segments: ISegment[];
}

export class Parser implements IParser {
  public readonly raw: string;
  public readonly segments: ISegment[];

  private _delimiters: IDelimiters;
  public get delimiters(): IDelimiters {
    return this._delimiters;
  }

  constructor(message: string) {
    this.raw = message;
    this.segments = [];
    this._delimiters = {} as any;

    // Setup
    this.setupDelimiters();
    this.setupSegments();
  }

  public toJson() {
    const response = {} as any;
    this.segments.forEach((segment) => {
      response[segment.name] = segment.toJson();
    });
    return response;
  }

  private setupDelimiters() {
    // TODO: Retrieve delimiters from the message itself.
    this._delimiters = DefaultDelimiters;
  }

  private setupSegments() {
    const splits = this.raw
      .split(this.delimiters.terminator)
      .filter((a) => a != "");
    splits.forEach((split) => {
      const segment = new Segment(split, {
        delimiters: this.delimiters,
      });
      this.segments.push(segment);
    });
  }
}
