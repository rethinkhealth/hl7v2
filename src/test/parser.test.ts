import * as fs from "fs";
import * as path from "path";

import { Parser } from "../parser";

describe("HL7v2 Parser", () => {
  it("should store the original message", () => {
    // Given
    const message = fs
      .readFileSync(path.join(__dirname, "../samples/siu_s12.txt"))
      .toString();

    // When
    const parser = new Parser(message);

    // Then
    expect(parser.raw).toEqual(message);
  });

  it("should retrieve the segments", () => {
    // Given
    const message = fs
      .readFileSync(path.join(__dirname, "../samples/siu_s12.txt"))
      .toString();

    // When
    const parser = new Parser(message);

    // Then
    expect(parser.segments.length).toEqual(8);
  });

  it("should return a JSON object", () => {
    // Given
    const message = fs
      .readFileSync(path.join(__dirname, "../samples/siu_s12.txt"))
      .toString();

    // When
    const parser = new Parser(message);

    // Then
    expect(parser.toJson()).toMatchSnapshot();
  });
});
