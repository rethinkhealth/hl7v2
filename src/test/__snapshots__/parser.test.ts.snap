// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HL7v2 Parser should return a JSON object 1`] = `
{
  "AIG": {
    "1": "AIG",
    "2": "1",
    "3": "",
    "4": {
      "1": "PULLEN",
      "2": "Pullen, Jeri",
    },
    "5": {
      "1": "P",
      "2": "",
      "3": "CSI",
    },
  },
  "AIL": {
    "1": "AIL",
    "2": "1",
    "3": "",
    "4": {
      "1": "MainOffi",
      "2": "",
      "3": "",
      "4": "MainOffi",
      "5": "",
      "6": "",
      "7": "",
      "8": "",
      "9": "Healthmatics Clinic",
    },
    "5": {
      "1": "",
      "2": "Healthmatics Clinic",
      "3": "CSI",
    },
  },
  "AIP": {
    "1": "AIP",
    "10": "15",
    "11": {
      "1": "m",
      "2": "Minutes",
    },
    "2": "1",
    "3": "",
    "4": {
      "1": "JPULLEN1",
      "2": "Pullen",
      "3": "Jeri",
      "4": "",
      "5": "",
      "6": "",
      "7": "",
      "8": "",
      "9": "&F12456&UPIN",
    },
    "5": {
      "1": "D",
      "2": "Pullen, Jeri",
    },
    "6": "",
    "7": "20190423140000",
    "8": "",
    "9": "",
  },
  "MSH": {
    "1": "MSH",
    "10": "8907-45",
    "11": "P",
    "12": "2.3",
    "13": "",
    "14": "",
    "15": "NE",
    "16": "NE",
    "2": {
      "1": {
        "1": "",
        "2": "",
      },
      "2": "\\&",
    },
    "3": "Ntierprise",
    "4": "Ntierprise Clinic",
    "5": "Healthmatics EHR",
    "6": "Healthmatics Clinic",
    "7": "20190423114154",
    "8": "",
    "9": {
      "1": "SIU",
      "2": "S12",
    },
  },
  "PID": {
    "1": "PID",
    "10": "",
    "11": "",
    "12": {
      "1": "007 Soho Lane",
      "2": "",
      "3": "Cary",
      "4": "NC",
      "5": "27511",
    },
    "13": "",
    "14": {
      "1": {
        "1": "(919)007-0007",
        "2": "",
        "3": "",
        "4": "",
        "5": "",
        "6": "919",
        "7": "0070007",
      },
      "2": {
        "1": "(777)707-0707",
        "2": "",
        "3": "CP",
        "4": "",
        "5": "",
        "6": "777",
        "7": "7070707",
      },
      "3": {
        "1": "",
        "2": "NET",
        "3": "X.400",
        "4": "007@BritishSecretService.com",
      },
    },
    "15": {
      "1": "(919)851-6177 X007",
      "2": "",
      "3": "",
      "4": "",
      "5": "",
      "6": "919",
      "7": "8516177",
      "8": "007",
    },
    "16": "",
    "17": "M",
    "18": "",
    "19": "150",
    "2": "1",
    "20": "007-00-0007",
    "21": "",
    "22": "",
    "23": "",
    "24": "",
    "25": "",
    "26": "",
    "27": "",
    "28": "",
    "29": "",
    "3": "",
    "30": "",
    "31": "N",
    "4": "150",
    "5": "",
    "6": {
      "1": "Bond",
      "2": "James",
      "3": "",
      "4": "007",
    },
    "7": "",
    "8": "19770920",
    "9": "M",
  },
  "PV1": {
    "1": "PV1",
    "10": {
      "1": "WEEDER",
      "2": "Weeder, M.D.",
      "3": "Dana",
      "4": "N",
      "5": "",
      "6": "",
      "7": "",
      "8": "",
      "9": "&W22630&UPIN",
    },
    "11": "",
    "12": "",
    "13": "",
    "14": "",
    "15": "",
    "16": "",
    "17": "N",
    "18": "",
    "19": "",
    "2": "1",
    "20": "",
    "21": "M",
    "3": "R",
    "4": "",
    "5": "",
    "6": "",
    "7": "",
    "8": {
      "1": "MRYAN",
      "2": "Ryan",
      "3": "Mark",
      "4": "S",
      "5": "phd",
      "6": "",
      "7": "",
      "8": "",
      "9": "&MR1127&UPIN",
    },
    "9": "",
  },
  "RGS": {
    "1": "RGS",
    "2": "1",
    "3": "A",
  },
  "SCH": {
    "1": "SCH",
    "10": "15",
    "11": "m",
    "12": {
      "1": "",
      "2": "",
      "3": "15",
      "4": "20190423140000",
      "5": "20190423141500",
    },
    "13": "",
    "14": "",
    "15": "",
    "16": "",
    "17": {
      "1": "mdrxmbyr",
      "2": "Byrne",
      "3": "Misty",
    },
    "18": "",
    "19": "",
    "2": "1209",
    "20": "",
    "21": {
      "1": "mdrxmbyr",
      "2": "Byrne",
      "3": "Misty",
    },
    "22": "",
    "23": "",
    "24": "",
    "25": "",
    "26": {
      "1": "Scheduled",
      "2": "",
      "3": "CSI",
    },
    "3": "13030",
    "4": "",
    "5": "",
    "6": "1209",
    "7": {
      "1": "OV15",
      "2": "OFFICE VISIT-15",
      "3": "CSI",
      "4": "N",
    },
    "8": {
      "1": "OFFICE VISIT-15",
      "2": "OFFICE VISIT-15 -",
    },
    "9": "OV15",
  },
}
`;
